internal/providers/nscloud/registry.go

```diff
@@ -7,36 +7,36 @@ package nscloud
 import (
 	"context"
 	"encoding/json"
-	"fmt"
-	"io"
-	"net/http"
-	"net/url"
+	"errors"
+	"io/fs"
+	"os"
+	"path/filepath"
 	"strings"
 
 	"github.com/google/go-containerregistry/pkg/authn"
-	"github.com/google/go-containerregistry/pkg/name"
+
 	"namespacelabs.dev/foundation/internal/artifacts/oci"
 	"namespacelabs.dev/foundation/internal/artifacts/registry"
-	"namespacelabs.dev/foundation/internal/auth"
 	"namespacelabs.dev/foundation/internal/compute"
 	"namespacelabs.dev/foundation/internal/fnerrors"
 	"namespacelabs.dev/foundation/internal/providers/nscloud/api"
+	"namespacelabs.dev/foundation/internal/workspace/dirs"
 	"namespacelabs.dev/foundation/std/cfg"
 	"namespacelabs.dev/foundation/std/tasks"
 )
 
-var DefaultKeychain oci.Keychain = defaultKeychain{}
+const (
+	registryConfigFile = "registry.json"
+)
 
-const loginEndpoint = "login.namespace.so/token"
+var DefaultKeychain oci.Keychain = defaultKeychain{}
 
 type nscloudRegistry struct {
 	clusterID string
 	registry  *api.ImageRegistry
 }
 
-const registryAddr = "registry-fgfo23t6gn9jd834s36g.prod-metal.namespacelabs.nscloud.dev"
-
-func RegisterRegistry() {
+func RegisterRegistry(imageRegistry *api.ImageRegistry) error {
```

Comment:
Hmm, passing the registry here feels odd, as we need to call this from many places then (as you do).
I would prefer, if we register a factory for the keychain that dynamically fetches the creds from checking the cluster ID, as we do for the registry.

From what I can tell, this would require a larger refactor to how the keychain is handled (making `cfg.Configuration` available) so let's check with @hugosantos before going down this path.

Reply:
As @n-g says, this information should be tied to the cluster, and should be looked up dynamically during runtime. And in fact never stored. If a user re-uses a nscloud cluster, this information will be fetched again from the VM service.

Reply:
Changed to request scopes tenant token to access registry.